File: javascript/day1/part2.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
});

let frequencyList = [];

rl.on("line", (input) => {
  frequencyList.push((Number.parseInt(input.trim())));
}).on("close", () => {
  let currentFrequency = 0;
  let frequenciesSeen = new Set();
  frequenciesSeen.add(currentFrequency);

  let currentIndex = 0;
  while (true) {
    currentFrequency += frequencyList[currentIndex];
    if (frequenciesSeen.has(currentFrequency)) {
      console.log(`Saw frequency ${currentFrequency} twice`);
      break;
    } else {
      currentIndex = (currentIndex + 1) % frequencyList.length;
      frequenciesSeen.add(currentFrequency);
    }
  }
})

File: python/day1/part2.py
import itertools
import sys


def main():
    input_list = [int(line.strip()) for line in sys.stdin]

    current_frequency = 0
    frequencies_seen = {0}

    for cycle_num, frequency in enumerate(itertools.cycle(input_list), 1):
        current_frequency += frequency
        if current_frequency in frequencies_seen:
            print(
                f"Saw frequency {current_frequency} twice on cycle {cycle_num}"
            )
            break
        else:
            frequencies_seen.add(current_frequency)


if __name__ == "__main__":
    main()
